src/build.rs

```diff
@@ -263,3 +265,323 @@ fn build_nasm(root: &Path, vendor_dir: &Path, out_dir: &Path, target_arch: &str,
     }
     n.compile_objects()
 }
+
+fn generate_ffi_mod() {
+    let out_dir = PathBuf::from(env::var_os("OUT_DIR").expect("outdir"));
+    let src_dir = env::current_dir().expect("failed to get current directory");
+
+    eprintln!("OUT_DIR={}", out_dir.display());
+    eprintln!("SRC_DIR={}", src_dir.display());
+    let ffi_input = src_dir.join("src").join("ffi.rs");
+    let ffi_src = std::fs::read_to_string(ffi_input.clone()).expect("failed to read src/ffi.rs");
+    println!("{}", format!("cargo:rerun-if-changed={}", ffi_input.display()));
+
+    // Parse src/ffi.rs generating Rust and C wrappers
+    let decls = parse_decls(&ffi_src);
+    let rust_wrapper = generate_rust_wrappers(&decls);
+    let cxx_wrapper = generate_cxx_wrappers(&decls);
```

Comment:
From a vector of the extern declarations, these two generate a `.rs` file containing the rust wrapper, and a `.cpp` file containing the C++ wrapper. The Rust wrapper is included by doing an `include!(...)` in the `lib.rs`, and the C++ wrapper is just compiled with the `cc` crate and linked. 

If you wanted, you could enable xLTO when an appropriate clang version is used, but you probably want to add support to the cc crate for that instead of putting the logic for this here.